### PR TITLE
Add a timeout feature

### PR DESCRIPTION
**Pull request by edeca**
Client will now be disconnected after a set period of time, which can be
adjusted.  This prevents one form of resource exhaustion when badly
behaving clients connect and never send any data.

This partially fixes issue #145

### CODE REVIEW

**Comment by asvetlov** [comment]
Per RFC there are several different timeouts depending on command.
Let's use http://www.attrs.org/en/stable/ data structure to specify them all instead of providing only the single value for everything.

**Comment by edeca** [comment]
The timeouts at the start of section 4.5.3.2 in RFC 2821 are for clients, not servers.

The relevant section appears to be:

> An SMTP server SHOULD have a timeout of at least 5 minutes while it is awaiting the next command from the sender.

**Comment by edeca** [comment]
Unsure why AppVeyor failed the second build.  The first one passed, the second commit only changes a typo of `es` to `is`.

**Comment by asvetlov** [comment]
Sorry.
@edeca you are right.

File: aiosmtpd/smtp.py [hunk 1]
```diff
@@ -182,6 +187,21 @@ def eof_received(self):
             return False
         return super().eof_received()
 
+    def _reset_timeout(self):
+        if self._timeout_handle:
```
Comment (asvetlov) [inline aiosmtpd/smtp.py]: Please use more explicit `is not None` check

File: aiosmtpd/tests/test_smtp.py [hunk 2]
```diff
@@ -1155,3 +1161,17 @@ def test_close_after_helo(self):
...
+    def setUp(self):
+        controller = TimeoutController(Sink)
+        controller.start()
+        self.addCleanup(controller.stop)
+        self.address = (controller.hostname, controller.port)
+
+    def test_timeout(self):
+        with SMTP(*self.address) as client:
+            code, response = client.ehlo('example.com')
+            time.sleep(2)
```
Comment (asvetlov) [inline aiosmtpd/tests/test_smtp.py]: 2 seconds is too long.
Please use `0.1` in configuration and `0.3` in the test.
They are typical values for asyncio test suite itself, should work fine for aiosmtp as well.

File: aiosmtpd/smtp.py (hunk 1 above)
Comment (edeca) [reply aiosmtpd/smtp.py]: I originally used `is not None` but this led to code coverage issues.  Will add this back and see whether I can find a meaningful test case.

**Comment by waynew** [comment]
@asvetlov Any reason not to go ahead and merge this? 👍 from me

**Comment by edeca** [comment]
I think all requested changes have been made - any chance of a merge and new release here?  

I currently have a cron job restarting my server every few hours (it handles approx. 500k messages per month).

**Comment by waynew** [comment]
@edeca If @asvetlov doesn't weigh in by Saturday morning I'll take another look and if I still agree with my past self, I'll go ahead ahead and merge this :slightly_smiling_face:

**Comment by waynew** [comment]
Still looks good 👍

**Comment by edeca** [comment]
Thanks @waynew, any chance of a tagged release and upload to PyPi at some point?

**Comment by waynew** [comment]
**tl;dr** - it's up to @warsaw 

I'm not sure what @warsaw 's plans are around that... I know there was some discussion about releasing after PyCon... 2017 :man_shrugging: 

I don't feel that I know enough about his plans for releases, etc. to give a good answer there. A big part of why I stepped up to help shepherd issues/PRs was because he asked for help, the other part is because I love SMTP, Python, and this project :)

I'm not sure if we have automated uploads to pypi or if that requires credentials, if it's the latter I know that I don't have them (yet?). If that's something that he also wants to allow some control over, I feel that I'd like to get some thoughts/feelings from Barry around the future, what sort of dreams/visions/roadmaps he has for aiosmtpd before I start making releases :slightly_smiling_face:

**Comment by warsaw** [comment]
@waynew I'm happy to hand over the keys to the PyPI kingdom for this package.  If you'd like to do releases, that would be fantastic.  I don't really have much time to do it these days.

**Comment by waynew** [comment]
@warsaw I'd be delighted! https://pypi.org/user/waynew/ is me on pypi, and I believe with the recent updates you can add me as a collaborator to the project (at least I recall reading about it on the distutils-sig and it looks like the UI supports that!)

**Comment by warsaw** [comment]
Done!  Thanks for all your great work on aiosmtpd @waynew !
